uwb-beacon-firmware/src/ekf.hpp

```diff
@@ -0,0 +1,91 @@
+#ifndef EKF_HPP
+#define EKF_HPP
+
+#include <Eigen/Dense>
+#include <utility>
+
+namespace EKF {
+/** Class responsible for the prediction part of a kalman filter.
+ *
+ * Its two templates arguments are the dimension of the state and the
+ * dimension of the control input.
+ */
+template <int N, int M> class Predictor {
+public:
+    typedef Eigen::Matrix<float, N, 1> State;
+    typedef Eigen::Matrix<float, N, N> Covariance;
+    typedef Eigen::Matrix<float, M, 1> Input;
+    typedef Eigen::Matrix<float, N, N> Jacobian;
+
+    Predictor(Covariance process_noise) : R(process_noise)
+    {
+    }
+
+    Predictor() : R(Covariance::Zero())
+    {
+    }
+
+    /** State update function. */
+    virtual State g(State state, Input input)
+    {
+        (void) input;
+        return state;
+    }
+
+    /** Returns the Jacobian of the state update function. */
+    virtual Jacobian G(State state, Input input)
+    {
+        (void) input;
+        (void) state;
+        return Jacobian::Identity();
+    }
+
+    std::pair<State, Covariance> predict(State mu, Covariance sigma, Input u)
```

Comment:
[trolling] Could have been a `operator()` override to make it a functor :p

Reply:
actually i did not know about operator() when I built it in cpp, the ekf.py module is implemented that way ;)